### DISCUSSION TITLE
How to update the drawer

### DISCUSSION THREAD

**Discussion by Yao-14** [Kitware/trame#221] [Q&A] [answered]
Hello, I am very interested in trame and have tried a lot on trame. 

Now I have a strange problem, how to update the drawer while updating the displayed data? Different data corresponds to different drawers.

**Comment by jourdain** [comment]
It really depend but I would suggest do the same thing as you would do with vue.js. This should be pretty straight forward since trame use vue.js for its client side rendering. 

If you can't get it managed dynamically via vue, you can also update the layout and flush it again. But I would not recommend that as it will force a rebuild of the UI on the client side and depending on how complex your app is, it might be somewhat costly.

But if you are not much of a developer, take the easy way. You can always optimize later on...

```python
def drawer_1():
    with layout:
        layout.drawer.clear()
        with layout.drawer:
            html.Div("Drawer 1")

def drawer_2():
    with layout:
        layout.drawer.clear()
        with layout.drawer:
            html.Div("Drawer 2")
```

**Comment by jourdain** [comment]
And the preferred way would be

```python

def drawer_1():
    state.drawer_content = "v1"

def drawer_2():
    state.drawer_content = "v2"

with layout:
   with layout.drawer:
       html.Div("Drawer 1", v_show="drawer_content === 'v1' ")
       html.Div("Drawer 2", v_show="drawer_content === 'v2' ")

```

**Comment by Yao-14** [comment]
Hi @jourdain , Thanks a lot for your help, but I'm still having some difficulties.

My purpose is that when I select a certain model in the toolbar or I input a certain model, the gittree in the drawer can change accordingly, but the current situation is that the gittree will not change.

![image](https://user-images.githubusercontent.com/89448797/224705330-9d74f4cd-819e-4bd3-8484-d1f56cc08e0e.png)

**Comment by Yao-14** [comment]
In the current situation, when I select a model in the toolbar, the model in the container will change accordingly, but the gittree will not change. I don’t know the specific reason. Can you help me solve this problem?

**Comment by jourdain** [comment]
When your model change, you just need to update the state of the pipeline for the git tree and the UI will update naturally. 
We have many apps that do exactly that and they work fine. 

Keep in mind that if you modify the same object, the state won't notice the difference and won't update the client. But you can [mark it dirty](https://trame.readthedocs.io/en/latest/core.state.html#trame_server.state.State.dirty) so it get flushed to the client when you want.

**Comment by Yao-14** [comment]
Thanks for your suggestion. could you give me a example of updating gitree?

**Reply by jourdain** [reply]
The same way you feed it with your data? What do you do? Might be easier to point you to what you do wrong.

Otherwise here is an example on how to update it. 

```python
# initial value
state.git_tree = [ ... ]; # v1

# update git tree
state.git_tree = [ ... ]; # v2
```

**Reply by Yao-14** [reply]
https://github.com/aristoteleo/spateo-viewer

**Reply by Yao-14** [reply]
This is a project I'm working on. Thanks for your help again.

**Reply by Yao-14** [reply]
My idea is that when I select different models in the toolbar, the gittree and cards in the drawer can be updated at the same time.

**Comment by jourdain** [comment, marked as answer]
So because of that [line](https://github.com/aristoteleo/spateo-viewer/blob/main/stviewer/ui/drawer.py#L52) all of [your gittrees](https://github.com/aristoteleo/spateo-viewer/blob/main/stviewer/app.py#L123-L128) are reacting to any change on `state.pipeline` and are displaying the exact same content (one is enough). One issue that I see is that you are forgetting to update that `state.pipeline` variable in [those functions](https://github.com/aristoteleo/spateo-viewer/blob/main/stviewer/ui/drawer.py#L26-L37) along with when you change the model you want to display.

Like I said previously, just call `state.pipeline = [ ... ]` every time you want the git tree to display something different.

**Reply by Yao-14** [reply]
Yes, you are quite right, thanks very much for your help. With your help I have solved the problem of gittree update.

**Comment by Yao-14** [comment]
I'm sorry to bother you again, I had problems again when updating the card, I don't know why there is no card under the new gittree?

**Reply by Yao-14** [reply]
https://github.com/aristoteleo/spateo-viewer/blob/main/stviewer/ui/drawer.py

**Comment by Yao-14** [comment]
I tried a method similar to gittree update, but there was never a very good result. Can you tell me the way to update card similar to gittree update?

**Comment by jourdain** [comment]
You need to organize your code in a way it make sense to you and what you aim to do. 
Trame is just giving you control to a Vue.js template with reactive data structure. Nothing more and nothing less.
